OADP-5901: Add design for DataProtectionTest CRD and Controller (#1706)
openshift/oadp-operator: docs/design/data-protection-test.md

Additionally, the design includes a mechanism to determine the S3-compatible vendor.

## Background

In several customer environments, backup/restore operations have experienced delays or stalls due to poor network connectivity, throttling at the cloud provider, or misconfigurations in BackupStorageLocation settings. Given the critical role of backup performance in disaster recovery, administrators need real-time, accurate metrics on how fast data uploads occur and how promptly snapshots are created.

This design addresses that need by measuring:

- **Upload speed:** How quickly data is transferred to object storage.
- **Snapshot performance:** How long a CSI VolumeSnapshot(s) take to become ReadyToUse.
- **S3 compatibility:** Identifying the specific S3 vendor by inspecting HTTP response headers from the storage endpoint.

## Goals

- Upload data and compute the speed in Mbps.
- Create a CSI VolumeSnapshot(s) and measure the time taken for it to become ready (reported as a duration).
- Identify the S3-compatible vendor using an HTTP call that inspects response headers.
- Expose all results—including upload speed, snapshot ready duration, and S3 vendor—in the CRD status.

> - Must gather will be updated th gather the DPT CR(s) for insights.

Reviewer: ```suggestion
- Must gather will be updated to gather the DPT CR(s) for insights.
```

## Non-Goals

- This design will not create or modify BackupStorageLocation entries in OADP.
- It will not implement download or latency tests, focusing solely on upload speed.
- Scheduling of recurring tests is not supported in the initial version.

> - Integration via DPA CR

Reviewer 2: Do you mean DPT CR here? We're not touching the DPA, right?

Author: No, thats why its a non-goal, right ?

Reviewer 2: @shubham-pampattiwar Oh yeah, I guess I got confused halfway through -- since the earlier ones said "will not", "not supported", etc. -- so by the time I got to this one, my brain had flipped it around. So this is good here.

## High-Level Design
Components involved and their responsibilities:

- **DataProtectionTest (DPT) CRD:**
    - **Spec:**

>         - **backupLocation:** Contains Velero-based backup storage configuration.

Reviewer 2: Is this just the name of the BSL, or does the user need to copy all of the BSL metadata?

Reviewer 2: Ahh, I see below. It's the full set of metadata. See suggestion in detailed design on this point.

        - **uploadSpeedTestConfig:** Test parameters for the upload (file size, test timeout).

>         - **CSIVolumeSnapshotTestConfig:** Test parameters for the CSI VolumeSnapshot test (snapshot class, source PVC name, and importantly the PVC namespace, plus snapshot timeout).

Reviewer 2: ```suggestion
        - **CSIVolumeSnapshotTestConfig:** Test parameters for the CSI VolumeSnapshot test (snapshot class, source PVC name and namespace for each PVC, plus snapshot timeout).
```

    - **Status:**

>         - **UploadTestStatus:** Groups upload speed (in Mbps), success flag, and error messages.

Reviewer 2: Should we also include upload size -- i.e. duration, size, as well as average speed?

Author: so size is already specified by the user in spec, so just duration ?

Author: Added duration

>         - **SnapshotTestStatus:** Groups the snapshot test results, reporting status and the duration taken for the snapshot to be ready.

Reviewer 2: ```suggestion
        - **SnapshotTestStatus:** Groups the snapshot test results, reporting status and the duration taken for the snapshot(s) to be ready.
```

Reviewer 2: Are we reporting per-snapshot duration? That probably makes more sense than total duration, since snapshots could become ready in parallel -- so the cumulative time may be misleading.

Author: yeah per snapshot duration.

>         - **S3Vendor:** Reports the detected S3 vendor string from vendor determination.

Reviewer 2: So I'm assuming this will be empty for non-s3 BSLs (azure, gcp).

Author: yes correct

- **DataProtectionTest Controller:**
    - Monitors DataProtectionTest CRs.
    - Extracts configuration from the backup location spec.
    - Determines the S3 vendor via an HTTP HEAD call.
    - Initializes the appropriate cloud provider using the CloudProvider interface.
    - Executes the upload test and, if enabled, the CSI VolumeSnapshot test.
    - Updates the CRD status with grouped results.
- **CloudProvider Interface:**
    - Defines an `UploadTest(ctx, config, bucket, fileSizeMB) (int64, error)` function.
    - AWS-specific implementation (S3Provider) is provided using the AWS SDK.
- **Vendor Determination Logic:**
    - A helper function performs an HTTP HEAD call to the **s3Url** and inspects headers (especially the `Server` header) to determine the vendor (e.g., "AWS", "MinIO", etc.).

## Detailed Design

#### Proposed DataProtectionTest CRD:

DataProtectionTest (DPT) CRD would look like:

```yaml
apiVersion: oadp.openshift.io/v1alpha1
kind: DataProtectionTest
metadata:
  name: my-data-protection-test
spec:

>   backupLocation:

Reviewer 2: Would it make sense to also have `spec.bslName` -- validate that only one of `bslName` and `backupLocation` can be present -- that way if the BSL is already defined in the cluster, they don't need to copy all of that metadata over -- they can just reference the BSL by name.

Author: Hmm, yeah we can do that, optional name, but then it would be either the BSLConfig or the BSLName, not both.

Author: Done, added a note and updated DPT CRD.

    provider: aws                # Cloud provider type (aws, azure, gcp)
    default: true
    objectStorage:
      bucket: sample-bucket
      prefix: velero
    config:
      region: us-east-1
      profile: "default"
      insecureSkipTLSVerify: "true"
      s3Url: "https://s3.amazonaws.com" # indicates s3 compatibility
    credential:
      name: cloud-credentials     # Secret for cloud credentials
      key: cloud
  uploadSpeedTestConfig:
    fileSize: "100MB"                # Size of file to upload for testing
    testTimeout: "60s"               # Maximum duration for upload test
  CSIVolumeSnapshotTestConfigs:
    - snapshotClassName: "csi-snapclass"
      volumeSnapshotSource:
        persistentVolumeClaimName: "db-pvc-1"
        persistentVolumeClaimNamespace: "app1"
      timeout: "60s"
    - snapshotClassName: "csi-snapclass"
      volumeSnapshotSource:
        persistentVolumeClaimName: "db-pvc-2"
        persistentVolumeClaimNamespace: "app1"
      timeout: "60s"
status:
  lastTested: "2024-10-08T10:00:00Z"
  uploadTest:

>     speedMbps: 55.3

Reviewer 2: Should we add totalMB and duration here too?

Author: size is already specified by the user, in the spec. So just duration ?

Author: Added duration.

Reviewer 2: Might be easier to see all 3 together -- and I don't know if it's possible for the transferred bytes to be different than specified, but it's not essential. Just adding duration is fine too.

# DataProtectionTest CRD and Controller Design

## Abstract

This design introduces the **DataProtectionTest** (DPT) CRD and its controller to evaluate data protection performance—including backup upload and volume snapshot readiness—in an OpenShift cluster.. It supports two tests:

> 1. **Upload Speed Test:** Measures the speed at which a dummy file is uploaded to cloud object storage using configuration from a BackupStorageLocation.

Reviewer: ```suggestion
1. **Upload Speed Test:** Measures the speed at which a test file of variable size is uploaded to cloud object storage using configuration from a BackupStorageLocation.
```

> 2. **CSI VolumeSnapshot Test:** Creates a VolumeSnapshot from a specified PVC and measures the time taken (as a duration) for the snapshot to become ready.

Reviewer: ```suggestion
2. **CSI VolumeSnapshot Test:** Creates  VolumeSnapshot(s) from a specified PVC and measures the time taken (as a duration) for the snapshot to become ready.
```

Additionally, the design includes a mechanism to determine the S3-compatible vendor.

## Background

In several customer environments, backup/restore operations have experienced delays or stalls due to poor network connectivity, throttling at the cloud provider, or misconfigurations in BackupStorageLocation settings. Given the critical role of backup performance in disaster recovery, administrators need real-time, accurate metrics on how fast data uploads occur and how promptly snapshots are created.

This design addresses that need by measuring:

- **Upload speed:** How quickly data is transferred to object storage.
- **Snapshot performance:** How long a CSI VolumeSnapshot takes to become ReadyToUse.
- **S3 compatibility:** Identifying the specific S3 vendor by inspecting HTTP response headers from the storage endpoint.

## Goals

- Upload a test file and compute the speed in Mbps.

> - Create a CSI VolumeSnapshot and measure the time taken for it to become ready (reported as a duration).

Reviewer: ```suggestion
- Create a CSI VolumeSnapshot(s) and measure the time taken for it to become ready (reported as a duration).
```

> - Identify the S3-compatible vendor using an HTTP call that inspects response headers.

Reviewer: Things to consider hitting:
https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketEncryption.html
https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketVersioning.html
others?

Author: we can do that, is this something we need along side the s3 vendor ?
cc: @sseago

Author: Done, updated the PR for bucket metadata support.

- Expose all results—including upload speed, snapshot ready duration, and S3 vendor—in the CRD status.
- Leverage existing BackupStorageLocation (from Velero/DPA) for configuration.

> 

Reviewer: Add Goal:  This tool is expected to be called and executed by the oadp-must-gather tool

Author: updated the PR

## Non-Goals

- This design will not create or modify BackupStorageLocation entries in OADP.
- It will not implement download or latency tests, focusing solely on upload speed.
- Scheduling of recurring tests is not supported in the initial version.

> 

Reviewer: - Non Goal:  This is not something that will be configured via the DPA

Author: updated the PR

## High-Level Design
Components involved and their responsibilities:

- **DataProtectionTest (DPT) CRD:**
    - **Spec:**
        - **backupLocation:** Contains Velero-based backup storage configuration.
        - **uploadSpeedTestConfig:** Test parameters for the upload (file size, test timeout).

>         - **CSIVolumeSnapshotTestConfig:** Test parameters for the CSI VolumeSnapshot test (snapshot class, source PVC name, and importantly the PVC namespace, plus snapshot timeout).

Reviewer 3: why not allow for auto create pvc of specified size to snapshot?
it is important to know how long it takes to provision pvc also for restore perf

Author: Discussed this during the sync all. The above was the general consensus.

}

```
5. Execute the CSI VolumeSnapshot Test (If enabled): Create a CSI VolumeSnapshot for a specified PVC and measure the time taken for it to be ready.
```go
// runSnapshotTest creates a CSI VolumeSnapshot from a PVC and measures the time until it's ReadyToUse.
func (r *DataProtectionTestReconciler) runSnapshotTest(ctx context.Context, dpt *oadpv1alpha1.DataProtectionTest) error {
    // Get PVC and snapshot class info
    // Parse snapshot readiness timeout duration
    // Create the VolumeSnapshot
    // Poll for ReadyToUse status within timeout
    // Check if snapshot is ReadyToUse
    // Success: capture duration
}

```
6. Update the Status of the DPT CR: Consolidate results from the upload test, snapshot test, and vendor detection, and update the CR status.

```mermaid
flowchart TD
  A[Start Reconciliation] --> B[Fetch DPT CR<br/>If not found, exit]
  B --> C[Determine S3 Vendor via inspecting HTTP request<br/>If s3Compatible is true]
  C --> D[Initialize Cloud Provider from BSL config]
  D --> E[Check if uploadSpeedTestConfig is present]
  E -->|Yes| F[Run Upload Speed Test<br/>Upload dummy file, measure speed]
  E -->|No| G[Skip Upload Test]
  F --> H
  G --> H

>   H[Check if CSI VolumeSnapshot Test is enabled] -->|Yes| I[Create VolumeSnapshot<br/>Poll until ReadyToUse]

Reviewer: If we executed the CSI test first, would we be able upload the CSI snapshot itself vs. a test file?  Does my thought make this too complicated for what we want to deliver atm?

  H -->|No| J[Skip Snapshot Test]
  I --> K[Calculate readyDuration]
  J --> K
  K --> L[Update DPT Status<br/>Set uploadSpeed, snapshot ready duration, vendor, lastTested]
  L --> M[End Reconciliation]

```

> #### Integration with DPA controller:

Reviewer: I think the integration with DPA can be removed based on team's discussion
